### ISSUE TITLE
adding/removing access control to a dataset results in additional links 

### ISSUE THREAD

**Issue by maaikelimper** [wmo-im/wis2box#775]
After adding removing access control to a dataset we end up with duplicate "observation"-buttons which appear to be due to links being added to the discovery metadata record whenever access-control is added and removed:

![image](https://github.com/user-attachments/assets/a3dc6888-2d2e-4715-a682-44c6a38eb7d3)

links section, note the duplicated links:

```
"links": [
{
"href": "http://dvila.wis2.training/oapi/collections/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=json",
"type": "application/json",
"name": "urn:wmo:md:ar-dvila:reco.surface-based-observations.synop",
"rel": "collection",
"title": "urn:wmo:md:ar-dvila:reco.surface-based-observations.synop"
},
{
"href": "mqtt://everyone:everyone@dvila.wis2.training:1883",
"type": "application/json",
"name": "origin/a/wis2/ar-dvila/data/recommended/weather/surface-based-observations/synop",
"rel": "items",
"channel": "origin/a/wis2/ar-dvila/data/recommended/weather/surface-based-observations/synop",
"title": "Notifications"
},
{
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop",
"type": "application/geo+json",
"name": "urn:wmo:md:ar-dvila:reco.surface-based-observations.synop",
"rel": "canonical",
"title": "urn:wmo:md:ar-dvila:reco.surface-based-observations.synop"
},
{
"type": "application/json",
"rel": "root",
"title": "The landing page of this server as JSON",
"href": "http://dvila.wis2.training/oapi?f=json"
},
{
"type": "text/html",
"rel": "root",
"title": "The landing page of this server as HTML",
"href": "http://dvila.wis2.training/oapi?f=html"
},
{
"rel": "self",
"type": "application/geo+json",
"title": "This document as JSON",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=json"
},
{
"rel": "alternate",
"type": "application/ld+json",
"title": "This document as RDF (JSON-LD)",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=jsonld"
},
{
"rel": "alternate",
"type": "text/html",
"title": "This document as HTML",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=html"
},
{
"rel": "collection",
"type": "application/json",
"title": "Discovery metadata",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata"
},
{
"type": "application/json",
"rel": "root",
"title": "The landing page of this server as JSON",
"href": "http://dvila.wis2.training/oapi?f=json"
},
{
"type": "text/html",
"rel": "root",
"title": "The landing page of this server as HTML",
"href": "http://dvila.wis2.training/oapi?f=html"
},
{
"rel": "self",
"type": "application/geo+json",
"title": "This document as JSON",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=json"
},
{
"rel": "alternate",
"type": "application/ld+json",
"title": "This document as RDF (JSON-LD)",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=jsonld"
},
{
"rel": "alternate",
"type": "text/html",
"title": "This document as HTML",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=html"
},
{
"rel": "collection",
"type": "application/json",
"title": "Discovery metadata",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata"
},
{
"type": "application/json",
"rel": "root",
"title": "The landing page of this server as JSON",
"href": "http://dvila.wis2.training/oapi?f=json"
},
{
"type": "text/html",
"rel": "root",
"title": "The landing page of this server as HTML",
"href": "http://dvila.wis2.training/oapi?f=html"
},
{
"rel": "self",
"type": "application/geo+json",
"title": "This document as JSON",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=json"
},
{
"rel": "alternate",
"type": "application/ld+json",
"title": "This document as RDF (JSON-LD)",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=jsonld"
},
{
"rel": "alternate",
"type": "text/html",
"title": "This document as HTML",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=html"
},
{
"rel": "collection",
"type": "application/json",
"title": "Discovery metadata",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata"
},
{
"type": "application/json",
"rel": "root",
"title": "The landing page of this server as JSON",
"href": "http://dvila.wis2.training/oapi?f=json"
},
{
"type": "text/html",
"rel": "root",
"title": "The landing page of this server as HTML",
"href": "http://dvila.wis2.training/oapi?f=html"
},
{
"rel": "self",
"type": "application/geo+json",
"title": "This document as JSON",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=json"
},
{
"rel": "alternate",
"type": "application/ld+json",
"title": "This document as RDF (JSON-LD)",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=jsonld"
},
{
"rel": "alternate",
"type": "text/html",
"title": "This document as HTML",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata/items/urn:wmo:md:ar-dvila:reco.surface-based-observations.synop?f=html"
},
{
"rel": "collection",
"type": "application/json",
"title": "Discovery metadata",
"href": "http://dvila.wis2.training/oapi/collections/discovery-metadata"
}
]
}
```

**Comment by maaikelimper** [comment]
@tomkralidis I suspect this issue is caused by the upsert_collection_item in these lines:
https://github.com/wmo-im/wis2box/blob/97bdaf75496974c9edfaae99cb104da3b8d6f134/wis2box-management/wis2box/auth.py#L195
https://github.com/wmo-im/wis2box/blob/97bdaf75496974c9edfaae99cb104da3b8d6f134/wis2box-management/wis2box/auth.py#L230

I guess the upsert_collection_item adds links instead of replacing them...

Could we delete/insert instead of upsert ?

**Comment by tomkralidis** [comment]
2024-11-21: @tomkralidis to apply fix (and review/inspect in the context of #807).
